### ISSUE TITLE
Unable to link against this for unittests + coverage

### ISSUE THREAD

**Issue by SerialVelocity** [vibe-d/vibe.d#767]
Hey, didn't know whether this is related to dub or vibe.d but when I try to compile my unittests + coverage with `dub test :subproject --coverage` it fails miserably with:

```
build type unittest-cov.
Compiling using dmd...
Linking...
Undefined symbols for architecture x86_64:
  "___coverage", referenced from:
      _D4vibe4core14connectionpool59__T14ConnectionPoolTC4vibe2db5redis5redis15RedisConnectionZ14ConnectionPool6__ctorMFNaNbNfDFZC4vibe2db5redis5redis15RedisConnectionZC4vibe4core14connectionpool59__T14ConnectionPoolTC4vibe2db5redis5redis15RedisConnectionZ14ConnectionPool in libvibe-d.a(redis.o)
      _D4vibe4core14connectionpool59__T14ConnectionPoolTC4vibe2db5redis5redis15RedisConnectionZ14ConnectionPool14lockConnectionMFZS4vibe4core14connectionpool61__T16LockedConnectionTC4vibe2db5redis5redis15RedisConnectionZ16LockedConnection in libvibe-d.a(redis.o)
      _D4vibe4core14connectionpool59__T14ConnectionPoolTC4vibe2db5redis5redis15RedisConnectionZ14ConnectionPool14lockConnectionMFZS4vibe4core14connectionpool61__T16LockedConnectionTC4vibe2db5redis5redis15RedisConnectionZ16LockedConnection12__dgliteral1MFNaNbNfZAya in libvibe-d.a(redis.o)
      _D4vibe4core14connectionpool59__T14ConnectionPoolTC4vibe2db5redis5redis15RedisConnectionZ14ConnectionPool14lockConnectionMFZS4vibe4core14connectionpool61__T16LockedConnectionTC4vibe2db5redis5redis15RedisConnectionZ16LockedConnection12__dgliteral2MFNbNfZm in libvibe-d.a(redis.o)
      _D4vibe4core14connectionpool59__T14ConnectionPoolTC4vibe2db5redis5redis15RedisConnectionZ14ConnectionPool14lockConnectionMFZS4vibe4core14connectionpool61__T16LockedConnectionTC4vibe2db5redis5redis15RedisConnectionZ16LockedConnection12__dgliteral3MFNaNbNfZm in libvibe-d.a(redis.o)
      _D4vibe4core14connectionpool59__T14ConnectionPoolTC4vibe2db5redis5redis15RedisConnectionZ14ConnectionPool14lockConnectionMFZS4vibe4core14connectionpool61__T16LockedConnectionTC4vibe2db5redis5redis15RedisConnectionZ16LockedConnection12__dgliteral4MFNaNbNfZAya in libvibe-d.a(redis.o)
      _D4vibe4core14connectionpool59__T14ConnectionPoolTC4vibe2db5redis5redis15RedisConnectionZ14ConnectionPool14lockConnectionMFZS4vibe4core14connectionpool61__T16LockedConnectionTC4vibe2db5redis5redis15RedisConnectionZ16LockedConnection12__dgliteral5MFNaNbNfZm in libvibe-d.a(redis.o)
      ...
ld: symbol(s) not found for architecture x86_64
```

**Comment by gedaiu** [comment]
there is a fix for this bug? I can reproduce it too...

**Comment by denizzzka** [comment]
Coverage test still isn't builds:

```
build type unittest-cov.
Compiling using dmd...
Linking...
../.dub/packages/vibe-d-0.7.23/libvibe-d.a(task.o): In function `_D4vibe4core4sync125__T13performLockedS62_D4vibe4core4task12MessageQueue5clearMFZ9__lambda1MFNaNbNiNfZvTC4vibe4core4sync22InterruptibleTaskMutexZ13performLockedMFNfC4vibe4core4sync22InterruptibleTaskMutexZv':
/home/denizzz/dianna2/../.dub/packages/vibe-d-0.7.23/source/vibe/core/sync.d:97: undefined reference to `__coverage'
/home/denizzz/dianna2/../.dub/packages/vibe-d-0.7.23/source/vibe/core/sync.d:99: undefined reference to `__coverage'
/home/denizzz/dianna2/../.dub/packages/vibe-d-0.7.23/source/vibe/core/sync.d:101: undefined reference to `__coverage'
/home/denizzz/dianna2/../.dub/packages/vibe-d-0.7.23/source/vibe/core/sync.d:101: undefined reference to `__coverage'
../.dub/packages/vibe-d-0.7.23/libvibe-d.a(task.o): In function `_D4vibe4core4sync158__T13performLockedS95_D4vibe4core4task12MessageQueue4sendMFS3std7variant18__T8VariantNVmi32Z8VariantNZ9__lambda2MFZvTC4vibe4core4sync22InterruptibleTaskMutexZ13performLockedMFC4vibe4core4sync22InterruptibleTaskMutexZv':
/home/denizzz/dianna2/../.dub/packages/vibe-d-0.7.23/source/vibe/core/sync.d:97: undefined reference to `__coverage'
../.dub/packages/vibe-d-0.7.23/libvibe-d.a(task.o):/home/denizzz/dianna2/../.dub/packages/vibe-d-0.7.23/source/vibe/core/sync.d:99: more undefined references to `__coverage' follow
collect2: error: ld returned 1 exit status
--- errorlevel 1
FAIL .dub/build/production-unittest-cov-linux.posix-x86_64-dmd_2067-554146DD9026E6DD2B5919533044B66B/ dianna2 executable
Error executing command test: dmd failed with exit code 1.
```

**Comment by luismarques** [comment]
I also stumbled into this issue. A basic vibe.d-based project with `dub test --coverage` complains:

```
Linking...
Undefined symbols for architecture x86_64:
  "___coverage", referenced from:
      _D4vibe4core4sync146__T13performLockedS74_D4vibe2db5redis5redis19RedisSubscriberImpl5bstopMFZ4implMFZ9__lambda1MFZvTC4vibe4core4sync31InterruptibleRecursiveTaskMutexZ13performLockedMFC4vibe4core4sync31InterruptibleRecursiveTaskMutexZv in libvibe-d.a(redis.o)
      _D4vibe4core4sync149__T13performLockedS86_D4vibe2db5redis5redis19RedisSubscriberImpl4stopMFZ4implMFZ9__lambda1MFZ9__lambda1MFZvTC4vibe4core4sync22InterruptibleTaskMutexZ13performLockedMFC4vibe4core4sync22InterruptibleTaskMutexZv in libvibe-d.a(redis.o)
      _D4vibe4core4sync145__T13performLockedS73_D4vibe2db5redis5redis19RedisSubscriberImpl4stopMFZ4implMFZ9__lambda1MFZvTC4vibe4core4sync31InterruptibleRecursiveTaskMutexZ13performLockedMFC4vibe4core4sync31InterruptibleRecursiveTaskMutexZv in libvibe-d.a(redis.o)
      _D4vibe4core4sync159__T13performLockedS96_D4vibe2db5redis5redis19RedisSubscriberImpl9subscribeMFMAAyaX4implMFZ9__lambda1MFZ9__lambda1MFZvTC4vibe4core4sync22InterruptibleTaskMutexZ13performLockedMFC4vibe4core4sync22InterruptibleTaskMutexZv in libvibe-d.a(redis.o)
      _D4vibe4core4sync155__T13performLockedS83_D4vibe2db5redis5redis19RedisSubscriberImpl9subscribeMFMAAyaX4implMFZ9__lambda1MFZvTC4vibe4core4sync31InterruptibleRecursiveTaskMutexZ13performLockedMFC4vibe4core4sync31InterruptibleRecursiveTaskMutexZv in libvibe-d.a(redis.o)
      _D4vibe4core4sync162__T13performLockedS99_D4vibe2db5redis5redis19RedisSubscriberImpl11unsubscribeMFMAAyaX4implMFZ9__lambda1MFZ9__lambda1MFZvTC4vibe4core4sync22InterruptibleTaskMutexZ13performLockedMFC4vibe4core4sync22InterruptibleTaskMutexZv in libvibe-d.a(redis.o)
      _D4vibe4core4sync158__T13performLockedS86_D4vibe2db5redis5redis19RedisSubscriberImpl11unsubscribeMFMAAyaX4implMFZ9__lambda1MFZvTC4vibe4core4sync31InterruptibleRecursiveTaskMutexZ13performLockedMFC4vibe4core4sync31InterruptibleRecursiveTaskMutexZv in libvibe-d.a(redis.o)
      ...
```

**Comment by s-ludwig** [comment]
I have no idea what to change on vibe.d's side, but `dub test --coverage --combined` works for me as a workaround in the meantime. I'm also not sure if this is a general issue w.r.t static libraries in how the coverage analysis is implemented, or if it's something that can be easily fixed in DMD.

**Comment by luismarques** [comment]
Yes, I was using the `dub test --coverage --combined` that I had seen suggested here, but having to find out about that is a bad experience that we should try to solve before it leaves a sour taste in the mouth of new D/dub/vibe.d users :-). If you are not sure how to solve this, how should we proceed? Should we ask on the D forums for help?

**Comment by s-ludwig** [comment]
I think this needs to go to the D bugtracker. I've started a dustmite run, but since it needs to rebuild everything at each iteration, it's extremely slow. It could also happen that it reduces to something that misses more symbols apart from `___coverage', but let's see. Manually creating a static library didn't reproduce the issue for me.

**Comment by s-ludwig** [comment]
DMD bug: https://issues.dlang.org/show_bug.cgi?id=13742

**Comment by luismarques** [comment]
Thanks, Sönke.

**Comment by Geod24** [comment]
The upstream bug has been fixed as part of DMD v2.078.1 (2018-01-22).
